tests/test_writer.py

```diff
@@ -105,81 +107,159 @@ def test_write_int_float_columns(self, tmp_path):
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_booleans, validate="madng")  # booleans are MAD-NG feature
+        assert write_location.is_file()
+
+        new = read_tfs(write_location)
+        assert_tfs_frame_equal(_tfs_dataframe_booleans, new, check_exact=False)  # float precision can be an issue
+
+    def test_tfs_write_read_with_complex(self, _tfs_dataframe_complex, tmp_path):
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_complex, validate="madng")  # booleans are MAD-NG feature
+        assert write_location.is_file()
+
+        new = read_tfs(write_location)
+        assert_tfs_frame_equal(_tfs_dataframe_complex, new, check_exact=False)  # float precision can be an issue
+
+    def test_tfs_write_with_nil_in_headers(self, _tfs_dataframe, tmp_path):
+        df = _tfs_dataframe
+        df.headers["WRITETONIL"] = None
+
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, df, validate="madng")
+
+        written = write_location.read_text()
+        assert "WRITETONIL" in written
+        assert "%n" in written
+        assert "nil" in written
+
+    def test_tfs_write_read_madng_like(self, _tfs_dataframe_madng, tmp_path):
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_madng, validate="madng")  # booleans are MAD-NG feature
+        assert write_location.is_file()
+
+        new = read_tfs(write_location)
+        assert_tfs_frame_equal(_tfs_dataframe_madng, new, check_exact=False)  # float precision can be an issue
+
+    @pytest.mark.skipif(sys.platform == "win32", reason="MAD-NG not available on Windows")
+    def test_tfs_write_madng_compatible_is_read_by_madng(self, _tfs_dataframe_madng, tmp_path):
+        from pymadng import MAD
+
+        write_location = tmp_path / "test.tfs"
+        write_tfs(write_location, _tfs_dataframe_madng, validate="madng")  # has all MAD-NG features
+        assert write_location.is_file()
+
+        # Now we read the file with pymadng and check there is not error from MADNG
+        # In case of error, pymadng will print the output to stdout/stderr. We need
+        # to ask to receive back from MAD-NG which would raise (and fail the test)
+        # if the loading did not go properly (our file is not accepted by MAD-NG).
+        madng = MAD(debug=True)
+        madng.protected_send(f"mtbl = mtable:read('{str(write_location.absolute())}')")
+
+        # If the loading went fine and we `.recv()` since we haven't sent anything it would
+        # just hang. So we first send some dummy and then call `.recv()`. Now if there was an
+        # error loading it raises a `RuntimeError` and the test fails, otherwise it returns a
+        # np.int32(1) which we don't care about, and the test concludes successfully.
+        madng.send("py:send(1)")
+        madng.recv()
```

Comment:
You don't check the contents after this??

Reply:
Would love to. How would I go at doing this?

Reply:
This might be useful: https://pymadng.readthedocs.io/en/latest/dataframes.html
My usage is here: https://github.com/pylhc/omc3/blob/6a9f229c3cea24c91c2666e0a15bc2200af5077b/tests/inputs/lhc_rdts/MAD_helpers.py#L286

Reply:
Wonderful, thank you!